src/wallet/asyncrpcoperation_common.cpp

```diff
@@ -40,3 +41,101 @@ std::pair<CTransaction, UniValue> SignSendRawTransaction(UniValue obj, std::opti
 
     return std::make_pair(tx, sendResult);
 }
+
+void ThrowInputSelectionError(const InputSelectionError& err) {
+    std::visit(match {
+        [](const AddressResolutionError& err) {
+            switch (err) {
+                case AddressResolutionError::SproutSpendNotPermitted:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "Sending from the Sprout shielded pool to the Sapling "
+                        "shielded pool is not enabled by default because it will "
+                        "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
+                        "Resubmit with the `privacyPolicy` parameter set to `AllowRevealedAmounts` "
+                        "or weaker if you wish to allow this transaction to proceed anyway.");
+                case AddressResolutionError::SproutRecipientNotPermitted:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "Sending funds into the Sprout pool is no longer supported.");
+                case AddressResolutionError::TransparentRecipientNotPermitted:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "This transaction would have transparent recipients, which is not "
+                        "enabled by default because it will publicly reveal transaction "
+                        "recipients and amounts. THIS MAY AFFECT YOUR PRIVACY. Resubmit "
+                        "with the `privacyPolicy` parameter set to `AllowRevealedRecipients` "
+                        "or weaker if you wish to allow this transaction to proceed anyway.");
+                case AddressResolutionError::InsufficientSaplingFunds:
+                    throw JSONRPCError(
+                        RPC_INVALID_PARAMETER,
+                        "Sending from the Sapling shielded pool to the Orchard "
+                        "shielded pool is not enabled by default because it will "
+                        "publicly reveal the transaction amount. THIS MAY AFFECT YOUR PRIVACY. "
+                        "Resubmit with the `privacyPolicy` parameter set to `AllowRevealedAmounts` "
+                        "or weaker if you wish to allow this transaction to proceed anyway.");
```

Comment:
Also, there is no corresponding error for sending Orchard to Sapling.

Reply:
This still needs to be fixed.

Reply:
Right, the error text here should be about sending _to_ Sapling. And the actual Sapling->Orchard case is handled under `UnifiedAddressResolutionError`.

Currently some of these messages are coarser than I’d like. There are two PRs later in this WalletTxBuilder DAG that improve on the situation, but there’s a balance between validating the tx at the end (in `GetRequiredPrivacyPolicy`) which can effectively look at the value balance and say “there are visible amounts moving Orchard to Sapling” and earlier checks that can add the context of “because there weren’t enough Sapling funds in the provided inputs.”